### PR TITLE
feat: add `Cruijff` PDF

### PR DESCRIPTION
**Pull request by ikrommyd**
Adds `Cruijff` PDF as unnormalized density
Related to https://github.com/zfit/zfit/issues/512

## Proposed Changes

  - Add `Cruijff` PDF as unnormalized density

## Tests added

  - `tests/test_pdf_cruijff.py`

## Checklist

 - [x] change approved
 - [x] implementation finished
 - [x] correct namespace imported
 - [x] tests added
 - [x] CHANGELOG updated

### CODE REVIEW

**Comment by ikrommyd** [comment]
@jonas-eschle FYI this is ready for review.

File: tests/test_pdf_cruijff.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    cruijff = zphys.pdf.Cruijff(mu=mu, sigmal=sigmal, alphal=alphal, sigmar=sigmar, alphar=alphar, obs=obs, norm=False)
+    return cruijff, obs
+
+
+def test_cruijff_pdf():
+    # Test PDF here
+    cruijff, _ = create_cruijff(
+        mu=mu_true, sigmal=sigmal_true, alphal=alphal_true, sigmar=sigmar_true, alphar=alphar_true, limits=(50, 130)
+    )
+    assert zfit.run(cruijff.pdf(90.0)) == pytest.approx(
```
Comment (jonas-eschle) [inline tests/test_pdf_cruijff.py]:
```suggestion
    assert cruijff.pdf(90.0).numpy() == pytest.approx(
```
`zfit.run` is quite outdated, I know that there are still places where it's used and it also works, but it just calls the `.numpy()` method. It's a relict from TF1

File: tests/test_pdf_cruijff.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    )
+    assert cruijff.pdf(tf.range(50.0, 130, 10_000)) <= cruijff.pdf(90.0)
+
+
+def test_cruihff_integral():
+    # Test CDF and integral here
+    cruijff, obs = create_cruijff(
+        mu=mu_true, sigmal=sigmal_true, alphal=alphal_true, sigmar=sigmar_true, alphar=alphar_true, limits=(50, 130)
+    )
+    full_interval_numeric = zfit.run(cruijff.numeric_integrate(obs, norm_range=False))
```
Comment (jonas-eschle) [inline tests/test_pdf_cruijff.py]:
```suggestion
    full_interval_numeric = cruijff.numeric_integrate(obs, norm=False).numpy()
```
Same with zfit.run, also, use `norm` whenever, `norm_range` is the old name

File: tests/test_pdf_cruijff.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    true_integral = 67.71494
+    numba_stats_full_integral = integrate.quad(
+        cruijff_numba.density, 50, 130, args=(alphal_true, alphar_true, mu_true, sigmal_true, sigmar_true)
+    )[0]
+    assert full_interval_numeric == pytest.approx(true_integral, 1e-7)
+    assert full_interval_numeric == pytest.approx(numba_stats_full_integral, 1e-7)
+
+    numeric_integral = zfit.run(cruijff.numeric_integrate(limits=(80, 100), norm_range=False))
```
Comment (jonas-eschle) [inline tests/test_pdf_cruijff.py]:
```suggestion
    numeric_integral = cruijff.numeric_integrate(limits=(80, 100), norm=False).numpy()
```

File: zfit_physics/models/pdf_cruijff.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        `tf.Tensor`: The calculated PDF values.
+
+    Notes:
+        Implementation from https://arxiv.org/abs/1005.4087
+    """
+    cond = znp.less(x, mu)
+
+    func = znp.where(
+        cond,
+        znp.exp(-0.5 * znp.square((x - mu) / sigmal) / (1 + alphal * znp.square((x - mu) / sigmal))),
```
Comment (jonas-eschle) [inline zfit_physics/models/pdf_cruijff.py]: Factor out the znp.square? It should be ideally the same in compiled mode, but in eager mode, it's surely slower

File: tests/test_pdf_cruijff.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import zfit
+from numba_stats import cruijff as cruijff_numba
+from scipy import integrate
+
+# Important, do the imports below
+from zfit.core.testing import tester
+
+import zfit_physics as zphys
+
+# specify globals here. Do NOT add any TensorFlow but just pure python
```
Comment (jonas-eschle) [inline tests/test_pdf_cruijff.py]:
```suggestion
```
Can be removed once done

File: tests/test_pdf_cruijff.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
+"""Tests for Cruijff PDF."""
+
+import numpy as np
+import pytest
+import tensorflow as tf
+import zfit
+from numba_stats import cruijff as cruijff_numba
+from scipy import integrate
+
+# Important, do the imports below
```
Comment (jonas-eschle) [inline tests/test_pdf_cruijff.py]: Same, can be removed once done
```suggestion
```

File: zfit_physics/models/pdf_cruijff.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        alphal: ztyping.ParamTypeInput,
+        sigmar: ztyping.ParamTypeInput,
+        alphar: ztyping.ParamTypeInput,
+        obs: ztyping.ObsTypeInput,
+        *,
+        extended: Optional[ztyping.ExtendedInputType] = False,
+        norm: Optional[ztyping.NormInputType] = None,
+        name: str = "Cruijff",
+    ):
+        """Cruijff PDF.
```
Comment (jonas-eschle) [inline zfit_physics/models/pdf_cruijff.py]:
```suggestion
        """Cruijff PDF, a Gaussian with two width, left and right, and non-Gaussian tails.
```

File: zfit_physics/models/pdf_cruijff.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+        Implementation from https://arxiv.org/abs/1005.4087
```
Comment (jonas-eschle) [inline zfit_physics/models/pdf_cruijff.py]:
```suggestion
        Implementation from https://arxiv.org/abs/1005.4087, citation 22, and http://hdl.handle.net/1794/9022, Equation IV.3
```
I would suggest to use this reference, as it's nicely written down, the Belle II papers have it as a citation point? Weirdest thing I've ever seen. But we can keep both

**Review by jonas-eschle — CHANGES_REQUESTED** [review]
Looks good overall, minor things, but other than that it's fine. A pity that no analytic integral is available, but fine, the numeric integrator catches it well. I'll approve, b

File: zfit_physics/models/pdf_cruijff.py (hunk 4 above)
Comment (ikrommyd) [reply zfit_physics/models/pdf_cruijff.py]: I'm not sure what that means. What equation form do you have in mind?

**Comment by ikrommyd** [comment]
> Looks good overall, minor things, but other than that it's fine. A pity that no analytic integral is available, but fine, the numeric integrator catches it well. I'll approve, b

I think there is just no anti-derivative. I tried with some integral calculators and got nowhere.

File: zfit_physics/models/pdf_cruijff.py (hunk 4 above)
Comment (jonas-eschle) [reply zfit_physics/models/pdf_cruijff.py]: I mean, doing something like... and as I wanted to give you an example, I just rewrote it quickly, it's more steps than I initially thought (even optimizes the traced mode too :D)

File: zfit_physics/models/pdf_cruijff.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        Implementation from https://arxiv.org/abs/1005.4087, citation 22, and http://hdl.handle.net/1794/9022, Equation IV.3
+    """
+    cond = znp.less(x, mu)
+
+    func = znp.where(
+        cond,
+        znp.exp(-0.5 * znp.square((x - mu) / sigmal) / (1 + alphal * znp.square((x - mu) / sigmal))),
+        znp.exp(-0.5 * znp.square((x - mu) / sigmar) / (1 + alphar * znp.square((x - mu) / sigmar))),
+    )
+    return func
```
Comment (jonas-eschle) [inline zfit_physics/models/pdf_cruijff.py]:
```suggestion
# compute only once (in graph this _may_ be optimized anyways, but surely not in eager)
xminmu = (x - mu)
tleft = znp.square(xminmu / sigmal)
tright = znp.square(xminmu / sigmar)
exponent = znp.where(
    cond,
    tleft / (1 + alphal * tleft),
    tright / (1 + alphar * tright),
)
value = znp.exp(-0.5 * exponent)
    return value
```
(also, I renamed func-> value, also `out` or anything is fine, but `func` is often used as a callable, absolutely minor detail)

**Review by jonas-eschle — APPROVED** [review]
I've added the suggestion to refactor. I think it makes sense? If so, just apply the suggestion and the PR is approved

**Comment by ikrommyd** [comment]
@jonas-eschle there was incorrect indentation in your suggestion but I fixed it. All good now, tests pass. Will merge as soon as ci completes.

**Comment by jonas-eschle** [comment]
> @jonas-eschle there was incorrect indentation in your suggestion but I fixed it. All good now, tests pass.

yep, just noticed, thanks for the fix

**Pull request merged**